layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+func (ds IPFIXDataSet) Len() int {
+	return int(ds.Header.Length)
+}
+
+func (ds *IPFIXDataSet) decodeFromBytes(data *[]byte, t IPFIXTemplateAccessor) error {
+	if len(*data) < ds.Len()-ds.Header.Len() {
+		*data = []byte{}
+		return fmt.Errorf("invalid data set length")
+	}
+
+	dataLen := ds.Len() - ds.Header.Len()
+	ds.Bytes = (*data)[:dataLen]
+
+	if t != nil {
+		// a Data Set can contains multiple Data Records
+		// we cannot know the exact length of a data record due to the variable
+		// length of certain field but we know the minimum. So until we have
+		// enough data for a record and we did not exceed the data set length,
+		// we can decode another record.
+		dataRead := 0
+		for len(*data) >= int(t.GetMinDataRecordLength()) && dataRead < dataLen {
+			dr := &IPFIXDataRecord{
+				ID:     ds.Header.ID,
+				Fields: []IPFIXField{},
+			}
+			n, err := dr.decodeFromBytes(data, t)
+			if err != nil {
+				return err
+			}
+			ds.Records = append(ds.Records, *dr)
+			dataRead += n
+		}
+
+		// remaining octet corresponding to padding see RFC 7011 section 3.1.1
+		padding := dataLen - dataRead
+		if padding > 0 {
+			for _, b := range (*data)[:padding] {
+				if b != 0x00 {
+					return fmt.Errorf("invalid data set padding value")
+				}
+			}
+			*data = (*data)[padding:]
+		}
+	} else {
+		*data = (*data)[dataLen:]
+	}
+
+	return nil
+}
+
+// IPFIXDataRecord is a Data Record (RFC 7011 section 3.4.3)
+//
+//      0                   1                   2                   3
+//      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Set ID = Template ID        |          Length               |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 1 - Field Value 1    |   Record 1 - Field Value 2    |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 1 - Field Value 3    |             ...               |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 2 - Field Value 1    |   Record 2 - Field Value 2    |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 2 - Field Value 3    |             ...               |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 3 - Field Value 1    |   Record 3 - Field Value 2    |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |   Record 3 - Field Value 3    |             ...               |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//     |              ...              |      Padding (optional)       |
+//     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IPFIXDataRecord struct {
+	// Template or Option template ID use to encode the record
+	ID IPFIXSetIDType
```

Comment:
this is already part of the set header?

Reply:
yes I copy it here as it can be a useful information when we will decode the record in the second step, i.e. if the template was unknown when we received the record.